test/suite/case.go

```diff
@@ -0,0 +1,208 @@
+
+	"github.com/futurehomeno/fimpgo"
+	"github.com/stretchr/testify/mock"
+
+	"github.com/futurehomeno/cliffhanger/router"
+	"github.com/futurehomeno/cliffhanger/task"
+)
+
+type Mock interface {
+	AssertExpectations(t mock.TestingT) bool
+}
+
+type Service interface {
+	Start() error
+	Stop() error
+}
+
+type Callback func(t *testing.T)
+
+func NewCase(name string) *Case {
+	return &Case{
+		Name: name,
+	}
+}
+
+type Case struct {
+	Name     string
+	Routing  []*router.Routing
+	Tasks    []*task.Task
+	Service  Service
+	Mocks    []Mock
+	Setup    Setup
+	TearDown []Callback
+	Nodes    []*Node
+}
+
+func (c *Case) WithName(name string) *Case {
+	c.Name = name
+
+	return c
+}
+
+func (c *Case) WithRouting(routing ...*router.Routing) *Case {
+	c.Routing = append(c.Routing, routing...)
+
+	return c
+}
+
+func (c *Case) WithTasks(tasks ...*task.Task) *Case {
+	c.Tasks = append(c.Tasks, tasks...)
+
+	return c
+}
+
+func (c *Case) WithService(service Service) *Case {
+	c.Service = service
+
+	return c
+}
+
+func (c *Case) WithMocks(mocks ...Mock) *Case {
+	c.Mocks = append(c.Mocks, mocks...)
+
+	return c
+}
+
+func (c *Case) WithSetup(setup Setup) *Case {
+	c.Setup = setup
+
+	return c
+}
+
+func (c *Case) WithNodes(nodes ...*Node) *Case {
+	c.Nodes = append(c.Nodes, nodes...)
+
+	return c
+}
+
+func (c *Case) WithTearDown(callbacks ...Callback) *Case {
+	c.TearDown = append(c.TearDown, callbacks...)
+
+	return c
+}
+
+func (c *Case) Run(t *testing.T, mqtt *fimpgo.MqttTransport) {
+	t.Helper()
+
+	c.init(t, mqtt)
+	defer c.tearDown(t)
+
+	for _, tn := range c.Nodes {
+		n := tn
+
+		t.Run(n.Name, func(t *testing.T) {
+			t.Helper()
+
+			n.Run(t, mqtt)
+		})
+	}
+
+	for _, m := range c.Mocks {
+		m.AssertExpectations(t)
+	}
+}
+
+func (c *Case) init(t *testing.T, mqtt *fimpgo.MqttTransport) {
+	t.Helper()
+
+	if c.Setup != nil {
+		c.Setup.apply(t, mqtt, c)
+	}
+
+	if len(c.Nodes) == 0 {
+		return
+	}
+
+	c.initRouting(mqtt)
+	c.initTasks()
+	c.initService()
+}
+
+func (c *Case) initRouting(mqtt *fimpgo.MqttTransport) {
+	if len(c.Routing) == 0 {
+		return
+	}
+
+	r := router.NewRouter(mqtt, "cliffhanger_test_case", c.Routing...)
+
+	initCallback := func(t *testing.T) {
+		err := r.Start()
+		if err != nil {
+			t.Fatalf("failed to start the router for the test case: %s", err)
+		}
+	}
+
+	c.Nodes[0].InitCallbacks = append(c.Nodes[0].InitCallbacks, initCallback)
+
+	tearDownCallback := func(t *testing.T) {
+		err := r.Stop()
+		if err != nil {
+			t.Fatalf("failed to stop the router for the test case: %s", err)
+		}
+	}
+
+	c.TearDown = append(c.TearDown, tearDownCallback)
+}
+
+func (c *Case) initTasks() {
+	if len(c.Tasks) == 0 {
+		return
+	}
+
+	taskManager := task.NewManager(c.Tasks...)
+
+	initCallback := func(t *testing.T) {
+		err := taskManager.Start()
+		if err != nil {
+			t.Fatalf("failed to start the task manager for the test case: %s", err)
+		}
+	}
+
+	c.Nodes[0].InitCallbacks = append(c.Nodes[0].InitCallbacks, initCallback)
```

Comment:
There is no other way of solving a race condition here without vastly modifying the entire structure. Simply put, the application or task manager must be started only AFTER the first set of expectations has been registered and is listened for. If we started a task manager or entire application BEFORE running a first test case node, there is a high chance that node expectations would be registered only after the first wave of messages were already sent to the MQTT. While this does not matter for command/response test scenarios, it is crucial for any time based task-manager related tests.

Reply:
If you don't want to touch the implementation, I'd at least expect a comment explaining the solution, because it looks a bit hacky at the first sight.

Reply:
It's not that I don't want to touch the implementation. "I don't like the idea" or "looks a bit hacky" is a personal feeling description, which is by definition correct, unless you're lying :-) However, If it is not obvious what should be changed, you need to provide a proposal of an alternative solution. Having init and teardown on the case level is not the solution, because of the reasons stated above. I've added a comment.